src/Compilers/CSharp/Portable/Binder/LocalBinderFactory.cs

```diff
@@ -707,32 +740,47 @@ private void AddToMap(CSharpSyntaxNode node, Binder binder)
             _map[node] = binder;
         }
 
+        private Binder GetBinderForPossibleEmbeddedStatement(StatementSyntax statement, Binder enclosing, out CSharpSyntaxNode embeddedScopeDisagnator)
+        {
+            switch (statement.Kind())
+            {
+                case SyntaxKind.LocalDeclarationStatement:
+                case SyntaxKind.LabeledStatement:
+                case SyntaxKind.LocalFunctionStatement:
+                // It is an error to have a declaration or a label in an embedded statement,
+                // but we still want to bind it.  
+
+                case SyntaxKind.ExpressionStatement:
+                case SyntaxKind.WhileStatement:
+                case SyntaxKind.DoStatement:
+                case SyntaxKind.LockStatement:
+                case SyntaxKind.IfStatement:
+                    Debug.Assert((object)_containingMemberOrLambda == enclosing.ContainingMemberOrLambda);
+                    embeddedScopeDisagnator = statement;
+                    return new EmbeddedStatementBinder(enclosing, statement);
+
+                case SyntaxKind.SwitchStatement:
+                    Debug.Assert((object)_containingMemberOrLambda == enclosing.ContainingMemberOrLambda);
+                    var switchStatement = (SwitchStatementSyntax)statement;
+                    embeddedScopeDisagnator = switchStatement.Expression;
+                    return new ExpressionVariableBinder(switchStatement.Expression, enclosing);
+
+                default:
+                    embeddedScopeDisagnator = null;
+                    return enclosing;
+            }
+        }
+
         private void VisitPossibleEmbeddedStatement(StatementSyntax statement, Binder enclosing)
         {
             if (statement != null)
```

Comment:
>if [](start = 12, length = 2)

Please add to the body of this method an explanation of the strategy - why it does things the way it does, and why it is necessary, why the embedded scope designator thing exists etc. None of this is clear without an explanation, and that explanation will be very useful to future readers of this code.